core-rust/core-api-server/src/core_api/metrics_layer.rs

```diff
@@ -0,0 +1,148 @@
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use std::{
+    sync::Arc,
+    task::{Context, Poll},
+    time::Instant,
+};
+
+use axum::{http::Request, response::Response};
+use futures_util::future::BoxFuture;
+use node_common::metrics::TakesMetricLabels;
+use tower::{Layer, Service};
+
+use super::metrics::CoreApiMetrics;
+
+#[derive(Debug, Clone)]
+pub struct MetricsLayer {
+    metrics: Arc<CoreApiMetrics>,
+}
+
+impl MetricsLayer {
+    pub fn new(metrics: Arc<CoreApiMetrics>) -> Self {
+        Self { metrics }
+    }
+}
+
+impl<S> Layer<S> for MetricsLayer {
+    type Service = MetricsService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        MetricsService {
+            inner,
+            metrics: self.metrics.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct MetricsService<S> {
+    inner: S,
+    metrics: Arc<CoreApiMetrics>,
+}
+
+impl<S, ReqBody> Service<Request<ReqBody>> for MetricsService<S>
+where
+    S: Service<Request<ReqBody>, Response = Response> + Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<ReqBody>) -> Self::Future {
+        let start = Instant::now();
+        let endpoint: String = request.uri().path().into();
+
+        let metrics = self.metrics.clone();
+        metrics.requests_pending.with_label(endpoint.clone()).inc();
+
+        let future = self.inner.call(request);
+
+        Box::pin(async move {
```

Comment:
Note: this `Box::pin` can be removed with a more cumbersome wrapping of `Future` which captures the context (metrics & start).